Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,199 @@
+-/
+noncomputable def preHilbert (d k : ℕ) : ℚ[X] :=
+  (d.factorial : ℚ)⁻¹ • (∏ i : Finset.range d, (X - (C (k : ℚ)) + (C (i : ℚ)) + 1))
+
+local notation "gFOSND" => greatestFactorOneSubNotDvd
+
+theorem preHilbert_eq_choose_sub_add (d k n : ℕ) (hkn : k ≤ n):
+    (preHilbert d k).eval (n : ℚ) = Nat.choose (n - k + d) d := by
+  rw [preHilbert]; simp only [Finset.univ_eq_attach, map_natCast, eval_smul, smul_eq_mul]
+  rw [eval_prod, @Finset.prod_attach ℕ ℚ _ (Finset.range d)
+    (fun j => eval n (X - (@Nat.cast ℚ[X] _ k) + (@Nat.cast ℚ[X] _ j) + 1))]
+  simp only [eval_add, eval_sub, eval_X, eval_one]
+  rw [Nat.add_choose, Nat.cast_div (Nat.factorial_mul_factorial_dvd_factorial_add _ _) (fun h => by
+    simp only [Nat.cast_mul, mul_eq_zero, Nat.cast_eq_zero] at h; exact Or.elim h
+      (Nat.factorial_ne_zero _) (Nat.factorial_ne_zero _)), Nat.cast_mul, div_mul_eq_div_div,
+      mul_comm, div_eq_mul_inv]
+  simp only [mul_eq_mul_right_iff, _root_.inv_eq_zero, Nat.cast_eq_zero]
+  · left
+    rw [← Nat.cast_div (Nat.factorial_dvd_factorial <| Nat.le_add_right (n - k) d) <| by
+      simp only [ne_eq, Nat.cast_eq_zero]; exact Nat.factorial_ne_zero (n - k),
+      ← Nat.ascFactorial_eq_div]
+    induction d with
+    | zero => simp only [Nat.zero_eq, Finset.range_zero, Finset.prod_empty, Nat.ascFactorial_zero,
+        Nat.cast_one]
+    | succ d hd => rw [Finset.prod_range_succ, hd, add_comm d, mul_comm, eval_natCast,
+        ← Nat.cast_sub hkn, eval_natCast, ← Nat.cast_add, ← Nat.cast_add_one, ← Nat.cast_mul,
+        ← Nat.succ_eq_one_add, Nat.ascFactorial_succ, add_assoc, add_comm d, ← add_assoc]
+
+/--
+Given `p : ℤ[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`.
+Look at `Polynomial.coeff_mul_invOneSubPow_eq_hilbert_eval`, which says
+that `PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d))` is equal to
+`(Polynomial.hilbert p d).eval (n : ℚ)` for any large enough `n : ℕ`.
+-/
+noncomputable def hilbert (p : ℤ[X]) (d : ℕ) : ℚ[X] :=
+  if h : p = 0 then 0
+  else if d + 1 ≤ p.rootMultiplicity 1 then 0
+  else ∑ i in Finset.range ((greatestFactorOneSubNotDvd p h).natDegree + 1),
+  ((greatestFactorOneSubNotDvd p h).coeff i) • preHilbert (d - p.rootMultiplicity 1) i
+
+/--
+Given `p : ℤ[X]` and `d : ℕ`. The key property of the Hilbert polynomial with respect to
+`p` and `d`, which says that for any term of `p * (@invOneSubPow ℤ _ d)` whose degree is
+large enough, its coefficient can be obtained by evaluating the Hilbert polynomial.
+-/
+theorem coeff_mul_invOneSubPow_eq_hilbert_eval (p : ℤ[X]) (d n : ℕ) (hn : p.natDegree < n) :
+    PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d)) = (hilbert p d).eval (n : ℚ) := by
+  rw [hilbert]; by_cases h : p = 0
+  · simp only [h, coe_zero, zero_mul, ↓reduceDIte, eval_zero, Int.cast_eq_zero]; rfl
+  · simp only [h, ↓reduceDIte, zsmul_eq_mul]
+    have coe_one_sub : (1 - X : ℤ[X]).ToPowerSeries = 1 - (PowerSeries.X : ℤ⟦X⟧) :=
```

Comment:
Hopefully this can be golfed! Maybe ask on Zulip in the Is there code for X channel?